### PR TITLE
T973: add basic node_exporter implementation

### PR DESCRIPTION
**Pull request by rebortg**
## Change Summary

Implement basic [node_exporter](https://github.com/prometheus/node_exporter) function

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)

https://vyos.dev/T973

## Related PR(s)

https://github.com/vyos/vyos-build/pull/757

## Component(s) name

node_exporter

## Proposed changes

add new feature node_exporter with vrf support.

## How to test

```
# use defaults:
set service monitoring node-exporter

# all config
set service monitoring node-exporter listen-address 10.10.10.1
set service monitoring node-exporter port 9999
set service monitoring node-exporter vrf mgmt
```

## Smoketest result

```
vyos@vyos:~$ /usr/libexec/vyos/tests/smoke/cli/test_service_monitoring_node-exporter.py
test_01_basic_config (__main__.TestMonitoringNodeExporter.test_01_basic_config) ... ok

----------------------------------------------------------------------
Ran 1 test in 4.832s

OK
```

## Checklist:

- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [x] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [x] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: interface-definitions/service_monitoring_node_exporter.xml.in [hunk 1]
```diff
@@ -0,0 +1,28 @@
+<?xml version="1.0"?>
+<interfaceDefinition>
+  <node name="service">
+    <children>
+      <node name="monitoring">
+        <children>
+          <node name="node-exporter" owner="${vyos_conf_scripts_dir}/service_monitoring_node-exporter.py">
+            <properties>
+              <help>node-exporter settings</help>
```
Comment (c-po) [inline interface-definitions/service_monitoring_node_exporter.xml.in]: Help should read: `Prometheus exporter for hardware and operating system metrics`

File: smoketest/scripts/cli/test_service_monitoring_node-exporter.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        )
+
+    @classmethod
+    def tearDownClass(cls):
+        cls.cli_delete(cls, ['interfaces', 'dummy', listen_if])
+        super(TestMonitoringNodeExporter, cls).tearDownClass()
+
+    def tearDown(self):
+        self.cli_delete(base_path)
+        self.cli_commit()
```
Comment (c-po) [inline smoketest/scripts/cli/test_service_monitoring_node-exporter.py:47]: Please test that process is no longer running after stopping the service.

File: src/conf_mode/service_monitoring_node-exporter.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        return None
+
+    # Render node_exporter and service_file
+    render_data = {
+        'deamon_args': list(),
+        'vrf_arg': '',
+    }
+    for address in config_data['listen_address']:
+        render_data['deamon_args'].append(
+            f"--web.listen-address={address}:{config_data['port']}"
```
Comment (c-po) [inline src/conf_mode/service_monitoring_node-exporter.py]: Any reason to not add this in the Jinja2 template itself?

File: src/conf_mode/service_monitoring_node-exporter.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        )
+
+    if 'vrf' in config_data:
+        render_data['vrf_arg'] = f"ip vrf exec {config_data['vrf']} "
```
Comment (c-po) [inline src/conf_mode/service_monitoring_node-exporter.py]: Any reason to not add this in the Jinja2 template itself like we do with SSH?

File: src/conf_mode/service_monitoring_node-exporter.py (hunk 3 above)
Comment (rebortg) [reply src/conf_mode/service_monitoring_node-exporter.py]: move it to service template file

File: interface-definitions/service_monitoring_node_exporter.xml.in (hunk 1 above)
Comment (rebortg) [reply interface-definitions/service_monitoring_node_exporter.xml.in]: correct help text

File: smoketest/scripts/cli/test_service_monitoring_node-exporter.py (hunk 2 above)
Comment (rebortg) [reply smoketest/scripts/cli/test_service_monitoring_node-exporter.py:47]: insert test for stopped service

File: src/conf_mode/service_monitoring_node-exporter.py (hunk 4 above)
Comment (rebortg) [reply src/conf_mode/service_monitoring_node-exporter.py]: move it to service template file

**Review by c-po — APPROVED** [review]

File: interface-definitions/service_monitoring_node_exporter.xml.in [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+        <children>
+          <node name="node-exporter" owner="${vyos_conf_scripts_dir}/service_monitoring_node-exporter.py">
+            <properties>
+              <help>Prometheus exporter for hardware and operating system metrics</help>
+              <priority>1280</priority>
+            </properties>
+            <children>
+              #include <include/listen-address.xml.i>
+              <leafNode name="listen-address">
+                <defaultValue>0.0.0.0</defaultValue>
```
Comment (frebib) [inline interface-definitions/service_monitoring_node_exporter.xml.in]: Probably should default to `::` so IPv6 scrapes work too

Comment (rebortg) [reply interface-definitions/service_monitoring_node_exporter.xml.in]: with 0.0.0.0 the exporter listen on *:9100 and so also on all IPv6 addresses:

```
vyos@vyos:~$ show configuration commands | grep node-exporter
set service monitoring node-exporter

vyos@vyos:~$ sudo ss -ltnp | grep 9100
LISTEN 0      4096               *:9100            *:*    users:(("node_exporter",pid=2336,fd=3))

vyos@vyos:~$ curl http://[2003:fa:af09:1d00:be24:11ff:fe99:28cb]:9100/metrics | head
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0# HELP go_gc_duration_seconds A summary of the pause duration of garbage collection cycles.
# TYPE go_gc_duration_seconds summary
go_gc_duration_seconds{quantile="0"} 2.7739e-05
go_gc_duration_seconds{quantile="0.25"} 2.7739e-05
go_gc_duration_seconds{quantile="0.5"} 6.2428e-05
go_gc_duration_seconds{quantile="0.75"} 6.2428e-05
go_gc_duration_seconds{quantile="1"} 6.2428e-05
go_gc_duration_seconds_sum 9.0167e-05
go_gc_duration_seconds_count 2
# HELP go_goroutines Number of goroutines that currently exist.
```

Comment (frebib) [reply interface-definitions/service_monitoring_node_exporter.xml.in]: Unintuitive but fair enough

Comment (dmbaturin) [reply interface-definitions/service_monitoring_node_exporter.xml.in]: If the daemon listens on all interfaces if listen address is not given, maybe just make it optional and don't give it any default?

File: data/templates/node_exporter/node_exporter.service.j2 [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+Description=Node Exporter
+Documentation=https://github.com/prometheus/node_exporter
+After=network.target
+
+[Service]
+{% if vrf is not vyos_defined %}
+User=node_exporter
+{% endif %}
+ExecStart={{ vrf_command }}/usr/sbin/node_exporter \
+{% for address in listen_address %}
```
Comment (c-po) [inline data/templates/node_exporter/node_exporter.service.j2]:
```suggestion
{% if listen_address is vyos_defined %}
  {% for address in listen_address %}
          --web.listen-address={{ address }}:{{ port }}
  {% endfor %}
{% endif %}
```

File: src/conf_mode/service_monitoring_node-exporter.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+
+    # we need to restart the service if e.g. the VRF name changed
+    systemd_action = 'reload-or-restart'
+    if 'restart_required' in config_data:
+        systemd_action = 'restart'
+
+    call(f'systemctl {systemd_action} {systemd_service}')
+
+    # Telegraf include custom rsyslog config changes
+    call('systemctl reload-or-restart rsyslog')
```
Comment (sever-sever) [inline src/conf_mode/service_monitoring_node-exporter.py]: Do you change the log format anywhere?
If not, we do not need to restart rsyslog, as it was required for telegraf only in some cases.

**Comment by sever-sever** [comment]
Probably requires to disable by default the service here https://github.com/vyos/vyos-build/blob/2b9303d0bd6b43bd92e2b2e9835a048877a63e11/data/live-build-config/hooks/live/18-enable-disable_services.chroot#L52

**Comment by rebortg** [comment]
> Probably requires to disable by default the service here https://github.com/vyos/vyos-build/blob/2b9303d0bd6b43bd92e2b2e9835a048877a63e11/data/live-build-config/hooks/live/18-enable-disable_services.chroot#L52

there is no service file in the node-exporter deb. The service file is create and delete by configure the feature

**Comment by sever-sever** [comment]
> there is no service file in the node-exporter deb. The service file is create and delete by configure the feature

For example, what will happen if I configure the service node exporter via VyOS CLI and reboot the router without saving?

File: smoketest/scripts/cli/test_service_monitoring_node-exporter.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
+#!/usr/bin/env python3
+#
+# Copyright (C) 2020-2024 VyOS maintainers and contributors
```
Comment (sever-sever) [inline smoketest/scripts/cli/test_service_monitoring_node-exporter.py]: Just 2024

File: src/conf_mode/service_monitoring_node-exporter.py [hunk 9]
```diff
@@ -0,0 +1,104 @@
+#!/usr/bin/env python3
+#
+# Copyright (C) 2021-2024 VyOS maintainers and contributors
```
Comment (sever-sever) [inline src/conf_mode/service_monitoring_node-exporter.py]: Just 2024

**Comment by rebortg** [comment]
> > there is no service file in the node-exporter deb. The service file is create and delete by configure the feature
> 
> For example, what will happen if I configure the service node exporter via VyOS CLI and reboot the router without saving?

good point. i tryed it.

the service file is present in `/etc/systemd/system/` but it is not enabled for autostart and so the service is not started

File: src/conf_mode/service_monitoring_node-exporter.py (hunk 7 above)
Comment (rebortg) [reply src/conf_mode/service_monitoring_node-exporter.py]: oh yes, stupid copy past error.

**Review by sever-sever — APPROVED** [review]

**Pull request merged**